### PR TITLE
[mlir] Add metadata to Diagnostic.

### PR DESCRIPTION
**Pull request by weiweichen**
Add metadata to Diagnostic. 

Motivation: we have a use case where we want to do some filtering in our customized Diagnostic Handler based on some customized info that is not `location` or `severity` or `diagnostic arguments` that are member variables of `Diagnostic`. Specifically, we want to add a unique ID to the `Diagnostic` for the handler to filter in a compiler pass that emits errors in async tasks with multithreading and the diagnostic handling is associated to the task.

This patch adds a field of `metadata` to `mlir::Diagnostics` as a general solution. `metadata` is of type `SmallVector<DiagnosticArgument, 0>` to save memory size and reuse existing `DiagnosticArgument` for metadata type.

### CODE REVIEW

File: mlir/include/mlir/IR/Diagnostics.h [hunk 1]
```diff
@@ -271,6 +271,9 @@ class Diagnostic {
     return failure();
   }
 
+  /// Returns the current list of diagnostic metadata.
+  SmallVector<DiagnosticArgument, 1>& getMetaData() { return metadata; }
```
Comment (joker-eph) [inline mlir/include/mlir/IR/Diagnostics.h:275]:
```suggestion
  SmallVectorImpl<DiagnosticArgument>& getMetaData() { return metadata; }
```

File: mlir/include/mlir/IR/Diagnostics.h [hunk 2]
```diff
@@ -290,6 +293,9 @@ class Diagnostic {
 
   /// A list of attached notes.
   NoteVector notes;
+
+  /// A list of metadata attached to this Diagnostic.
+  SmallVector<DiagnosticArgument, 1> metadata;
```
Comment (joker-eph) [inline mlir/include/mlir/IR/Diagnostics.h:298]:
```suggestion
  SmallVector<DiagnosticArgument, 0> metadata;
```

**Review by joker-eph — CHANGES_REQUESTED** [review]
Thanks for the contributions:

- Can you please provide some tests? These also serve as examples for the feature.
- Can you also elaborate on the use-cases in the PR description?

**Comment by weiweichen** [comment]
> Thanks for the contributions:
> 

> * Can you please provide some tests? These also serve as examples for the feature.

Yep! Added `mlir/test/IR/diagnostic-handler-metadata.mlir`

> * Can you also elaborate on the use-cases in the PR description?

Yes, updated the PR description with a bit details on the downstream use case we have that needs this upstream change.

Thank you very much for the feedback!

**Comment by weiweichen** [comment]
@joker-eph could you please take a look to see if there should be more changes with this PR? Thank you!

File: mlir/test/IR/diagnostic-handler-metadata.mlir [hunk 3]
```diff
@@ -0,0 +1,19 @@
+// RUN: mlir-opt %s -pass-pipeline="builtin.module(func.func(test-diagnostic-metadata))" -o - 2>&1 | FileCheck %s
+// This test verifies that diagnostic handler can filter the diagnostic whether to emit the errors.
+
+// CHECK-LABEL: Test 'test'
+// CHECK-NEXT: 8:3: error: test diagnostic metadata
+// CHECK-NOT: 13:3: error: test diagnostic metadata
```
Comment (joker-eph) [inline mlir/test/IR/diagnostic-handler-metadata.mlir:6]: Instead of using FileCheck, can't we use the `--verify-diagnostic` mode of mlir-opt?

**Review by joker-eph — COMMENTED** [review]
Thanks for adding the test!

Is there some documentation to update as well?

File: mlir/test/IR/diagnostic-handler-metadata.mlir (hunk 3 above)
Comment (weiweichen) [reply mlir/test/IR/diagnostic-handler-metadata.mlir]: Ah, yes, I was following [what `diagnostic-handler-filter.mlir` is done for checking the diagnostics](https://github.com/llvm/llvm-project/blob/2e57e6366677390110f5382894c8afeba8da7419/mlir/test/IR/diagnostic-handler-filter.mlir#L5-L6). 

`-verify-diagnostic` is a much better way, updated the test.

**Comment by weiweichen** [comment]
> Thanks for adding the test!
> 
> Is there some documentation to update as well?

Yep, added something to `Diagnostic.md` about `metadata`.

**Review by joker-eph — APPROVED** [review]

**Pull request merged**